resources/net.go

```diff
@@ -0,0 +1,671 @@
+	// messages left in the channel
+	ifaceChan := make(chan *dbus.Signal, 10) // iface event channel
+	defer close(ifaceChan)
+
+	bus.Signal(ifaceChan)
+	defer bus.RemoveSignal(ifaceChan)
+
+	// build the path to the configuration file
+	obj.unitFile = NetworkdUnitFileDir + IfacePrefix + obj.GetName() + NetworkdUnitFileExt
+
+	// watch the configuration file
+	recWatcher, err := recwatch.NewRecWatcher(obj.unitFile, false)
+	if err != nil {
+		return err
+	}
+	defer recWatcher.Close()
+
+	// notify engine that we're running
+	if err := obj.Running(); err != nil {
+		return err // bubble up a NACK...
+	}
+
+	var send = false
+	var exit *error
+
+	for {
+		select {
+		case event, ok := <-ifaceChan:
+			if !ok {
+				return fmt.Errorf("channel closed unexpectedly")
+			}
+			if obj.debug { // don't access event.Body if event.Error isn't nil
+				log.Printf("%s: Event(%s): %v", obj, event.Sender, event.Body)
+			}
+
+			send = true
+			obj.StateOK(false) // dirty
+
+		case event, ok := <-recWatcher.Events():
+			if !ok { // channel shutdown
+				return nil
+			}
+			if err := event.Error; err != nil {
+				return errwrap.Wrapf(err, "Unknown %s watcher error", obj)
+			}
+			if obj.debug { // don't access event.Body if event.Error isn't nil
+				log.Printf("%s: Event(%s): %v", obj, event.Body.Name, event.Body.Op)
+			}
+
+			send = true
+			obj.StateOK(false) // dirty
+
+		case event, ok := <-svcChan:
+			if !ok {
+				return fmt.Errorf("channel closed unexpectedly")
+			}
+			if obj.debug {
+				log.Printf("%s: Event(%s): %v", obj, event[NetworkdService].ActiveState, event)
+			}
+
+			send = true
+			obj.StateOK(false) // dirty
+
+		case err, ok := <-svcErr:
+			if !ok {
+				return fmt.Errorf("channel closed unexpectedly")
+			}
+			return errwrap.Wrapf(err, "unknown %s error", obj)
+
+		case event, ok := <-obj.Events():
+			if !ok {
+				return fmt.Errorf("channel closed unexpectedly")
+			}
+			if exit, send = obj.ReadEvent(event); exit != nil {
+				return *exit // exit
+			}
+		}
+
+		// do all our event sending all together to avoid duplicate msgs
+		if send {
+			send = false
+			obj.Event()
+		}
+	}
+}
+
+// ifaceCheckApply checks the state of the network device and performs the
+// appropriate actions to converge its state with the graph.
+func (obj *NetRes) ifaceCheckApply(apply bool) (checkOK bool, _ error) {
+	// check the interface state.
+	iface, err := net.InterfaceByName(obj.GetName())
+	if err != nil {
+		return false, errwrap.Wrapf(err, "error finding interface: %s", obj.GetName())
+	}
+
+	// check if the interface is up
+	ifaceUp := strings.Contains(iface.Flags.String(), "up")
+
+	// if the interface's state is correct, we're done
+	if obj.State == "down" && !ifaceUp {
+		return true, nil
+	}
+	// check iface addrs
+	addrs, err := iface.Addrs()
+	if err != nil {
+		return false, errwrap.Wrapf(err, "error getting addrs")
+	}
+	if obj.State == "up" && ifaceUp {
+		// error if we lose the connection and softfail is false
```

Comment:
I'd like to discuss this with you at some point, maybe before you review. This doesn't really work like it should. Instead I should be getting the `/org/freedesktop/network1/link/$foo` `OperationalState` property so I can detect when it's `no-carrier` to know when we've lost connectivity. But, I can't figure out how to get the property with the dbus api (need to input a systemd unit, nothing I've tried works, but I can get it in d-feet)

Reply:
![screenshot from 2017-12-23 08-48-33](https://user-images.githubusercontent.com/30417125/34320034-27f257ae-e7be-11e7-977e-84f500345259.png)